google/cloud/internal/time_utils_test.cc

```diff
@@ -24,46 +24,115 @@ inline namespace GOOGLE_CLOUD_CPP_NS {
+  // Note: These values can be computed with `date +%s --date="YYYY-MM-...Z"`
+  auto const min_proto = MakeProto(-62135596800, 0);
+  auto const max_proto = MakeProto(253402300799, 999999999);
+
+  auto const min_time = TimeFromString("0001-01-01T00:00:00Z").value();
+  auto const max_time =
+      TimeFromString("9999-12-31T23:59:59.999999999Z").value();
+  auto const contemporary =
+      TimeFromString("2020-07-21T11:33:03.123456789Z").value();
+
+  struct {
+    absl::Time t;
+    google::protobuf::Timestamp expected;
+  } round_trip_cases[] = {
+      {min_time, min_proto},
+      {epoch - sec(1), MakeProto(-1, 0)},
+      {epoch - sec(1), MakeProto(-1, 0)},
+      {epoch - sec(1) + ns(1), MakeProto(-1, 1)},
+      {epoch - ns(1), MakeProto(-1, 999999999)},
+      {epoch, MakeProto(0, 0)},
+      {epoch + ns(1), MakeProto(0, 1)},
+      {epoch + sec(1), MakeProto(1, 0)},
+      {epoch + sec(1) + ns(1), MakeProto(1, 1)},
+      {epoch + sec(123) + ns(456), MakeProto(123, 456)},
+      {contemporary, MakeProto(1595331183, 123456789)},
+      {max_time, max_proto},
+  };
+
+  for (auto const& tc : round_trip_cases) {
+    SCOPED_TRACE("Time: " + absl::FormatTime(tc.t));
+    auto const p = ToProtoTimestamp(tc.t);
+    EXPECT_THAT(p, IsProtoEqual(tc.expected));
+    auto const t = ToAbslTime(p);
+    EXPECT_EQ(t, tc.t);
+  }
+
+  // Tests that times before the min time are capped at the min time.
+  EXPECT_THAT(min_proto, IsProtoEqual(ToProtoTimestamp(absl::InfinitePast())));
+  EXPECT_THAT(min_proto, IsProtoEqual(ToProtoTimestamp(min_time - ns(1))));
+
+  // Tests that times after the max time are capped at the max time.
+  EXPECT_THAT(max_proto, IsProtoEqual(ToProtoTimestamp(max_time + ns(1))));
+  EXPECT_THAT(max_proto,
+              IsProtoEqual(ToProtoTimestamp(absl::InfiniteFuture())));
 }
 
-TEST(TimeUtils, ConvertAbslTimeToProtoTimestamp) {
-  google::protobuf::Timestamp expected;
-  expected.set_seconds(867);
-  expected.set_nanos(530900);
-  auto const t = absl::FromUnixSeconds(867) + absl::Nanoseconds(530900);
-  auto const actual = ToProtoTimestamp(t);
-  EXPECT_THAT(expected, IsProtoEqual(actual));
+TEST(TimeUtils, ChronoTime) {
+  // Define a few constants and lambdas to make the test cases below read
+  // nicely with minimal wrapping.
+  // Note: we use microseconds rather than nanoseconds here because we test on
+  // some platforms where the system clock only has microsecond resolution.
+  auto const epoch = std::chrono::system_clock::from_time_t(0);
+  auto const sec = [](std::int64_t n) { return std::chrono::seconds(n); };
+  auto const us = [](std::int64_t n) { return std::chrono::microseconds(n); };
+  auto const contemporary = absl::ToChronoTime(
+      TimeFromString("2020-07-21T11:33:03.123456000Z").value());
+  // Note: we do not test the proto min/max values here because we don't know
+  // if chrono's system_clock can represent that value (it likely can't).
```

Comment:
FWIW, it looks like we also assume `time_point` has microsecond resolution.

Reply:
Yes, that's what I meant in this comment. Updated comment to make it clearer.

Reply:
I'd say the range and resolution capabilities are different beasts, but OK.

Reply:
FWIW, you can probably static assert whether the `std::chrono` thing has at least microsecond resolution.

```cc
  bool system_clock_has_micros = std::ratio_greater_equal<
      std::micro, std::chrono::system_clock::duration::period>::value;
```

and if you can do that, then you can write a test at full resolution when available, and no-op when it is not...

Totally optional, I do not think it matters.

Reply:
@devbww I may not be following exactly what you're suggesting. If you have wording you'd like me to paste in, feel free to let me know. We do test on some platforms where the system_clock has microsecond resolution, which I mentioned in the comment on line 104, so that's why the test uses microsecond resolution. However, we also test on platforms w/ nanosecond resolution, and on those platforms, they can't represent the full proto min/max range, which is why I don't have a test for that case.

@coryan 
> Totally optional, I do not think it matters.

Thanks. I think that would add a bit of complexity for little additional value in our tests, so I think I'll skip it in this PR.